File: edulink/apps/internships/logbook_format.py
from datetime import date, datetime, timedelta


ATTACHMENT_LOGBOOK_FORMAT_VERSION = "edulink_attachment_logbook_v1"

ATTACHMENT_WEEKDAYS = [
    {"key": "MON", "label": "Monday", "offset": 0},
    {"key": "TUE", "label": "Tuesday", "offset": 1},
    {"key": "WED", "label": "Wednesday", "offset": 2},
    {"key": "THU", "label": "Thursday", "offset": 3},
    {"key": "FRI", "label": "Friday", "offset": 4},
]


def parse_iso_date(value: str) -> date:
    return datetime.strptime(value, "%Y-%m-%d").date()


def format_iso_date(value: date) -> str:
    return value.isoformat()


def get_week_start_date(value: date) -> date:
    return value - timedelta(days=value.weekday())


def get_required_week_start_dates(start_date: date | None, end_date: date | None) -> list[str]:
    if not start_date or not end_date or end_date < start_date:
        return []

    current = get_week_start_date(start_date)
    final_week = get_week_start_date(end_date)
    required_weeks = []

    while current <= final_week:
        required_weeks.append(format_iso_date(current))
        current += timedelta(days=7)

    return required_weeks


def get_week_ending_date(week_start_date: str) -> str:
    return format_iso_date(parse_iso_date(week_start_date) + timedelta(days=4))


def get_logbook_week_start(metadata: dict | None) -> str:
    normalized = normalize_logbook_metadata(metadata)
    return normalized.get("week_start_date", "")


def build_standard_daily_entries(*, week_start_date: str, entries: dict) -> list[dict]:
    start = parse_iso_date(week_start_date)
    return [
        {
            "day": day["key"],
            "label": day["label"],
            "date": format_iso_date(start + timedelta(days=day["offset"])),
            "description": (entries or {}).get(format_iso_date(start + timedelta(days=day["offset"])), ""),
        }
        for day in ATTACHMENT_WEEKDAYS
    ]


def normalize_logbook_metadata(metadata: dict | None) -> dict:
    metadata = metadata or {}
    week_start_date = metadata.get("week_start_date") or metadata.get("weekStartDate")
    entries = metadata.get("entries") or {}
    daily_entries = metadata.get("daily_entries")

    if not week_start_date and daily_entries:
        week_start_date = daily_entries[0].get("date")

    if not week_start_date:
        return {
            **metadata,
            "format_version": ATTACHMENT_LOGBOOK_FORMAT_VERSION,
            "logbook_type": "INDUSTRIAL_ATTACHMENT",
            "week_start_date": "",
            "week_ending_date": "",
            "weekly_summary": metadata.get("weekly_summary", ""),
            "daily_entries": daily_entries or [],
            "entries": entries,
        }

    if not daily_entries:
        daily_entries = build_standard_daily_entries(
            week_start_date=week_start_date,
            entries=entries,
        )

    return {
        **metadata,
        "format_version": ATTACHMENT_LOGBOOK_FORMAT_VERSION,
        "logbook_type": "INDUSTRIAL_ATTACHMENT",
        "week_start_date": week_start_date,
        "week_ending_date": metadata.get("week_ending_date") or get_week_ending_date(week_start_date),
        "weekly_summary": metadata.get("weekly_summary", ""),
        "daily_entries": daily_entries,
        "entries": entries,
    }

File: edulink-frontend/src/utils/logbookFormat.ts
export const ATTACHMENT_LOGBOOK_FORMAT_VERSION = 'edulink_attachment_logbook_v1';

export const ATTACHMENT_WEEKDAYS = [
  { key: 'MON', label: 'Monday', offset: 0 },
  { key: 'TUE', label: 'Tuesday', offset: 1 },
  { key: 'WED', label: 'Wednesday', offset: 2 },
  { key: 'THU', label: 'Thursday', offset: 3 },
  { key: 'FRI', label: 'Friday', offset: 4 },
] as const;

export type AttachmentWeekdayKey = typeof ATTACHMENT_WEEKDAYS[number]['key'];

export interface StandardDailyLogEntry {
  day: AttachmentWeekdayKey;
  label: string;
  date: string;
  description: string;
}

export interface StandardLogbookMetadata {
  format_version: typeof ATTACHMENT_LOGBOOK_FORMAT_VERSION;
  logbook_type: 'INDUSTRIAL_ATTACHMENT';
  week_start_date: string;
  week_ending_date: string;
  weekly_summary: string;
  daily_entries: StandardDailyLogEntry[];
  entries: Record<string, string>;
}

const parseLocalDate = (dateStr: string) => new Date(`${dateStr}T00:00:00`);

export const getLocalDateString = (date: Date) => {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
};

export const addDays = (dateStr: string, days: number) => {
  const date = parseLocalDate(dateStr);
  date.setDate(date.getDate() + days);
  return getLocalDateString(date);
};

export const getWeekEndingDate = (weekStartDate: string) => addDays(weekStartDate, 4);

export const buildStandardDailyEntries = (
  weekStartDate: string,
  entries: Record<string, string>
): StandardDailyLogEntry[] => ATTACHMENT_WEEKDAYS.map(day => {
  const date = addDays(weekStartDate, day.offset);
  return {
    day: day.key,
    label: day.label,
    date,
    description: entries[date] || '',
  };
});

export const buildStandardLogbookMetadata = ({
  weekStartDate,
  entries,
  weeklySummary,
}: {
  weekStartDate: string;
  entries: Record<string, string>;
  weeklySummary: string;
}): StandardLogbookMetadata => ({
  format_version: ATTACHMENT_LOGBOOK_FORMAT_VERSION,
  logbook_type: 'INDUSTRIAL_ATTACHMENT',
  week_start_date: weekStartDate,
  week_ending_date: getWeekEndingDate(weekStartDate),
  weekly_summary: weeklySummary.trim(),
  daily_entries: buildStandardDailyEntries(weekStartDate, entries),
  entries,
});

export const countCompletedStandardDays = (entries: Record<string, string>, weekStartDate: string) =>
  buildStandardDailyEntries(weekStartDate, entries).filter(entry => entry.description.trim()).length;
